app/streaming/video/ffmpeg-renderers/eglvid.cpp

```diff
@@ -0,0 +1,480 @@
+        glTexParameteri(GL_TEXTURE_EXTERNAL_OES, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+    }
+
+    GLenum err = glGetError();
+    if (err != GL_NO_ERROR) {
+        SDL_LogError(SDL_LOG_CATEGORY_APPLICATION, "OpenGL error: %d", err);
+        deinitialize();
+    }
+
+    return err == GL_NO_ERROR;
+}
+
+void EGLRenderer::deinitialize() {
+    if (m_context) {
+        if (m_shader_program) {
+            glDeleteProgram(m_shader_program);
+            m_shader_program = 0;
+        }
+        if (m_vao) {
+            glDeleteVertexArrays(1, &m_vao);
+            m_vao = 0;
+        }
+        if (m_egl_display) {
+            // EGL context should be handled by SDL
+            m_egl_display = nullptr;
+        }
+        SDL_GL_DeleteContext(m_context);
+        m_context = nullptr;
+    }
+}
+
+const float *EGLRenderer::getColorMatrix() {
+    /* The conversion matrices are shamelessly stolen from linux:
+     * drivers/media/platform/imx-pxp.c:pxp_setup_csc
+     */
+    static const float bt601_lim[] = {
+        1.1644f, 1.1644f, 1.1644f,
+        0.0f, -0.3917f, 2.0172f,
+        1.5960f, -0.8129f, 0.0f
+    };
+    static const float bt601_full[] = {
+        1.0f, 1.0f, 1.0f,
+        0.0f, -0.3441f, 1.7720f,
+        1.4020f, -0.7141f, 0.0f
+    };
+    static const float bt709_lim[] = {
+        1.1644f, 1.1644f, 1.1644f,
+        0.0f, -0.2132f, 2.1124f,
+        1.7927f, -0.5329f, 0.0f
+    };
+    static const float bt709_full[] = {
+        1.0f, 1.0f, 1.0f,
+        0.0f, -0.1873f, 1.8556f,
+        1.5748f, -0.4681f, 0.0f
+    };
+    static const float bt2020_lim[] = {
+        1.1644f, 1.1644f, 1.1644f,
+        0.0f, -0.1874f, 2.1418f,
+        1.6781f, -0.6505f, 0.0f
+    };
+    static const float bt2020_full[] = {
+        1.0f, 1.0f, 1.0f,
+        0.0f, -0.1646f, 1.8814f,
+        1.4746f, -0.5714f, 0.0f
+    };
+
+    switch (m_colorspace) {
+        case AVCOL_SPC_SMPTE170M:
+        case AVCOL_SPC_BT470BG:
+            SDL_LogInfo(SDL_LOG_CATEGORY_APPLICATION,
+                        "EGLRenderer: BT-601 pixels");
+            return m_color_full ? bt601_full : bt601_lim;
+        case AVCOL_SPC_BT709:
+            SDL_LogInfo(SDL_LOG_CATEGORY_APPLICATION,
+                        "EGLRenderer: BT-709 pixels");
+            return m_color_full ? bt709_full : bt709_lim;
+        case AVCOL_SPC_BT2020_NCL:
+        case AVCOL_SPC_BT2020_CL:
+            SDL_LogInfo(SDL_LOG_CATEGORY_APPLICATION,
+                        "EGLRenderer: BT-2020 pixels");
+            return m_color_full ? bt2020_full : bt2020_lim;
+    };
+    SDL_LogWarn(SDL_LOG_CATEGORY_APPLICATION,
+                "EGLRenderer: unknown color space: %d, falling back to BT-601",
+                m_colorspace);
+    return bt601_lim;
+}
+
+bool EGLRenderer::specialize() {
+    if (!compileShader())
+        return false;
+    if (m_vao)
+        glDeleteVertexArrays(1, &m_vao);
+
+    // XXX: Maybe we should keep the window ratio for the vertices
```

Comment:
What does this comment mean?

Reply:
It means that I should compute the ratio of the image I will display to keep the same ratio as the source, see https://github.com/moonlight-stream/moonlight-qt/pull/365#discussion_r420645039